### ISSUE TITLE
Custom Request Logging on AWS Lambda

### ISSUE THREAD

**Issue by rv-ekramer** [jovotech/jovo-framework-nodejs#94]
Hey guys,

I'm building a skill to help you find what streaming service you can watch a show and recommend shows based on another show.

How would I use something like `app.onRequest()` and `app.onResponse()`  (as listed in the changelog of 1.0.3 here https://github.com/jovotech/jovo-framework-nodejs/blob/master/CHANGELOG.md)  in a Lambda?

I know the default generated skill has a call to `if(app.isWebhook()){ ... }` and then outside of that block has a named export that makes a call to `app.handleLambda()` to handle the Lambda. How would I modify this to use  `app.onRequest()` and `app.onResponse()` for custom response logging?

**Comment by aswetlow** [comment]
You can use it in your /app/app.js
´´´javascript
app.onRequest(function(jovo) {
    const requestObject = jovo.getRequestObject();
}
´´´

**Comment by rv-ekramer** [comment]
@aswetlow I can use it in `app/app.js`? Where would I put it? Before the call to `app.setHandler(...)?`

So we'd have something like this?

```javascript
const { App } = require('jovo-framework');
const app = new App({});

app.onRequest(function(jovo) {
    console.log(jovo.getRequestObject());
}

app.onResponse(function(jovo) {
    console.log(jovo.getResponseObject());
}

app.setHandler({
     // ...
});
```

**Comment by aswetlow** [comment]
Exactly :+1:

**Comment by rv-ekramer** [comment]
Cool cool thanks!

**Comment by aswetlow** [comment]
Have you tried to delete node_modules and do a ```npm install``` again?

**Comment by rv-ekramer** [comment]
I’ll try that and let you know.

On Mar 6, 2018, at 5:51 PM, Alex <notifications@github.com<mailto:notifications@github.com>> wrote:


Have you tried to delete node_modules and do a npm install again?

**Comment by rv-ekramer** [comment]
No such luck unfortunately 😞 
<img width="1099" alt="screen shot 2018-03-07 at 2 52 05 pm" src="https://user-images.githubusercontent.com/31253814/37114797-4e34776a-2217-11e8-9cd4-82e1e232e135.png">

**Comment by rv-ekramer** [comment]
Tried `rm -rf node_modules` and `npm i` a few times. I have a `postinstall` script in my `package.json` that runs `npm install -g jovo-cli` too

**Comment by aswetlow** [comment]
Hm, it looks like a broken dependency (in the express module). Thought that a ````rm -rf node_module``` and ```npm i``` would fix it (like here: https://stackoverflow.com/questions/40976709/express-js-cannot-find-module-mime-types) 

How does the package.json look like?

**Comment by rv-ekramer** [comment]
```json
{
  "name": "tv-helper-one-oh-upgrade",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "postinstall": "npm install -g jovo-cli"
  },
  "dependencies": {
    "array-shuffle": "^1.0.1",
    "axios": "^0.18.0",
    "jovo-framework": "^1.0.4",
    "node-fetch": "^2.1.1"
  },
  "devDependencies": {
    "jovo": "0.0.1",
    "nodemon": "^1.17.1"
  },
  "author": "Red Ventures"
}
````

**Comment by rv-ekramer** [comment]
@aswetlow Is this issue specific to `1.0.4` of `jovo-framework`? It didn't happen on `1.0.0`?

**Comment by rv-ekramer** [comment]
Update `running npm install --save express body-parser` fixed the issue. Weird that I have to explicitly install them though.

**Comment by rv-ekramer** [comment]
When I add `jovo run --watch` as a dev script to my `package.json` all it does is say:

```bash
> tv-helper-one-oh-upgrade@1.0.0 dev /Users/ekramer/Desktop/voice-assistants/tv-helper
> jovo run --watch

I'm setting everything up
```

**Comment by aswetlow** [comment]
I'm wondering that jovo (0.0.1) is in your dev dependencies. I think it was the first version of the cli. What's the version of the cli?
```jovo -v```

**Comment by rv-ekramer** [comment]
`jovo -v` yields `Jovo CLI Version: 1.0.6`

**Comment by rv-ekramer** [comment]
Whats the diff between `jovo`, 'jovo-framework`, `jovo-framework-nodejs` and `jovo-cli`?

**Comment by rv-ekramer** [comment]
Earlier it felt like the `jovo` package was needed to run `jovo-cli`. Should i remove `jovo` from my package.json?

**Comment by rv-ekramer** [comment]
Also when running `jovo run --watch` during simple things like invoking the skill through "open tv-helper" or invoking an intent that simply returns a static text response (e.g. doesn't fetch data from an API) sometimes I get the following error. Sometimes it happens pretty and sometimes it hardly happens:

```bash
Unhandled Rejection at: Promise Promise {
  <rejected> { Error: connect ECONNREFUSED 127.0.0.1:3000
    at Object._errnoException (util.js:1022:11)
    at _exceptionWithHostPort (util.js:1044:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1182:14)
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 3000 } } reason: { Error: connect ECONNREFUSED 127.0.0.1:3000
    at Object._errnoException (util.js:1022:11)
    at _exceptionWithHostPort (util.js:1044:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1182:14)
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 3000 }
Server listening on port 3000!
```

But when it happens the Alexa says "The requested skill took too long to respond". Any ideas why this might be happening?

**Comment by aswetlow** [comment]
yes, remove it. 

jovo-framework: Cross platform framework that helps you to build voice apps for Alexa and Google Assistant.
jovo-cli: Command line tool that helps you to create/build and maintain Jovo projects easily.

**Comment by aswetlow** [comment]
For local development we create a lightweight Express server that handles all requests and responses. Somehow yours is not reachable all the time. 

When does it happen?

**Comment by rv-ekramer** [comment]
@aswetlow Running `npm uninstall --save jovo` fixed error I was getting when using `npm run dev` to run `jovo run --watch`

The unhandled promise rejection error usually happens during simple requests like invoking the app with "open tv helper" or when returning a static response without reaching to an API for anything, but I've found it also happens when trying to invoke an intent with an utterance that isn't explicitly stated in the sample utterances of the intent in the dev tools.

**Comment by rv-ekramer** [comment]
**Update:** It seems to be happening every other request regardless of the simplicity or the complexity of the routed intents logic

**Comment by jankoenig** [comment]
Hey @rv-ekramer maybe it's because the server is restarting and the request is coming in while it does so, hence not reachable? Could you see if it's happening without the `--watch`?

**Comment by pmontague** [comment]
Yo, got some clap / tweet love for me over in Europe?

https://medium.com/@pmontee/on-information-communications-technology-and-our-lives-cdfd3ef387cf?source=linkShare-ad5cf9451e38-1521399259

Sent while mobile.

**Comment by rv-ekramer** [comment]
Hey @jankoenig the dev server seems to be a little more reliable when just running `jovo run` without `--watch`. What would this mean? I feel like part of my problem is that from the terminal output, it appears that the server has finished re-starting from nodemon before it actually has, and maybe thats causing the issue.

**Comment by jankoenig** [comment]
Yes, somehow the "ignore db-folder" option doesn't seem to work, so it's restarting after every request. Will investigate. cc @aswetlow

**Comment by ekdevdes** [comment]
Ohhh ok. Got it. That kinda makes sense

**Comment by ekdevdes** [comment]
^ by the way this is still me. This is just my personal GitHub. The former is my work account.

**Comment by aswetlow** [comment]
Fixed! Please update your jovo-cli to 1.0.9

**Comment by rv-ekramer** [comment]
Awesome thanks!
